libsolidity/analysis/ConstantEvaluator.cpp

```diff
@@ -25,26 +25,99 @@
 
 #include <libsolidity/ast/AST.h>
 #include <libsolidity/ast/TypeProvider.h>
+
 #include <liblangutil/ErrorReporter.h>
 
+#include <libsolutil/Common.h>
+
 using namespace std;
 using namespace solidity;
 using namespace solidity::frontend;
 using namespace solidity::langutil;
 
 void ConstantEvaluator::endVisit(UnaryOperation const& _operation)
 {
-	auto sub = type(_operation.subExpression());
+	auto sub = evaluatedValue(_operation.subExpression());
 	if (sub)
-		setType(_operation, sub->unaryOperatorResult(_operation.getOperator()));
+		setValue(_operation, sub->unaryOperatorResult(_operation.getOperator()));
+}
+
+std::optional<rational> ConstantEvaluator::evaluateBinary(rational _left, rational _right, Token _operator, SourceLocation _location)
```

Comment:
Do we really need to uplicate this from Types.cpp?

Reply:
I was looking all around the source code and I think I only found two use cases for `Type::binaryOperatorResult` (and unary...); conversion test and actual constant evaluation.

The latter (constant evaluation) however is _only_ implemented in `RationalNumberType`, all the others remain with `Type::binaryOperatorResult` not doing any constant evaluation (but doing conversion ops).

I think it's best to factor our the constant evaluation part from `RationalNumberType` as it *only* happens there and I think it better fits the other class. If that is a wise choice, I'd like to get some feedback from others. So, I think we could remove the evaluation part in Types.cpp because it's only in this one class for binary/unary operations. What do you think?